app/decorators/service_provider_session.rb

```diff
@@ -139,7 +139,11 @@ def sp_ial
   end
 
   def requested_aal
-    sp_session[:aal_level_requested] || 1
+    AuthnContextResolver.new(
```

Comment:
Here the AAL level is ultimately used to determine the remember device cookie expiration:

https://github.com/18F/identity-idp/blob/4dd9359e8bf23aea6a83cec9c1454365973367f1/app/decorators/service_provider_session.rb#L91-L99

I think instead of copying the old value we could use `AuthnContextResolver` in the remember device cookie expiration computation:

```ruby
  def mfa_expiration_interval
    aal_1_expiration = IdentityConfig.store.remember_device_expiration_hours_aal_1.hours
    aal_2_expiration = IdentityConfig.store.remember_device_expiration_minutes_aal_2.minutes
    return aal_2_expiration if sp_aal > 1
    return aal_2_expiration if sp_ial > 1
    # return aal_2_expiration if requested_aal > 1
    return aal_2_expiration if resolved_authn_context_result.aal2?

    aal_1_expiration
  end
```

Of course we'll need to define `resolved_authn_context_result` here for that to work. Alternatively we could pass it in. If we pass it in the next 2 paragraphs become irrelevant.

I also noticed you are passing `nil` here for the `service_provider` arg. The service provider is available in this class as `sp`. Setting it will make sure we are getting results that include the SP defaults. If we do that well we can also remove those `sp_aal` and `sp_ial` checks in `#mfa_expiration_interval`

Finally, we probably want to make sure we are reading `vtr` and `acr_values` from `sp_session` instead of from `sp_request`. They shouldn't be different but I think we want to make sure we are looking at the same place as the application controller.

If we pass the `resolved_authn_context_result` we can still make the changes to remove those `sp_aal` and `sp_ial` checks in `#mfa_expiration_interval`. The result that comes from the application controller includes the SP defaults when it is setting `aal2?`.